avc/src/planner/planner.cpp

```diff
@@ -1,129 +1,199 @@
-                    best_path_angle = angle;
-                }
-
-				/*if (std::abs(angle) < std::abs(best_path_angle)) {
-					best_path_speed = speed;
-					best_path_angle = angle;
-				} else if (speed > best_path_speed) {
-					best_path_speed = speed;
-					best_path_angle = angle;
-				}*/
-			}
-		}
-	}
-	rr_platform::speedPtr speedMSG(new rr_platform::speed);
-	rr_platform::steeringPtr steerMSG(new rr_platform::steering);
-	speedMSG->speed = best_path_speed;
-	steerMSG->angle = best_path_angle;
-	speed_pub.publish(speedMSG);
-	steer_pub.publish(steerMSG);
-	nav_msgs::Path path;
-	for(double t = 0; t < best_path_speed * TIMESTEP; t += DISTANCE_INCREMENT) {
-		pose step = calculateStep(0, 0, 0, best_path_speed, best_path_angle, t / best_path_speed);
-		geometry_msgs::PoseStamped p;
-		p.pose.position.x = step.x;
-		p.pose.position.y = step.y;
-		path.poses.push_back(p);
-	}
-	path.header.frame_id = "map";
-	//ROS_INFO_STREAM(path.poses.size());
-	path_pub.publish(path);
+    pcl::PCLPointCloud2 pcl_pc2;
+    pcl_conversions::toPCL(*map, pcl_pc2);
+    pcl::PointCloud<pcl::PointXYZ>::Ptr cloud(new pcl::PointCloud<pcl::PointXYZ>);
+    pcl::fromPCLPointCloud2(pcl_pc2, *cloud);
+    if(cloud->empty()) {
+        ROS_WARN("Point cloud is empty");
+        rr_platform::speedPtr speedMSG(new rr_platform::speed);
+        rr_platform::steeringPtr steerMSG(new rr_platform::steering);
+        speedMSG->speed = 0.5; //proceed with caution; E-Kill if necessary
+        steerMSG->angle = 0;
+        speed_pub.publish(speedMSG);
+        steer_pub.publish(steerMSG);
+        return;
+    }
+
+    pcl::KdTreeFLANN<pcl::PointXYZ> kdtree;
+    kdtree.setInputCloud(cloud);
+
+    //build paths and evaluate their weights
+    double weightTotal, cost, angle;
+    int nPathPoses;
+    weightTotal = nPathPoses = 0;
+    vector<double> weightedSteeringRaw;
+    for(int i = 0; i < PATH_ITERATIONS; i++) {
+        vector<double> steerPath;
+        for(int s = 0; s < PATH_STAGES; s++) {
+            angle = steeringSample();
+            for(double t = 0; t <= PATH_STAGE_TIME; t += TIME_INCREMENT) {
+                steerPath.push_back(angle);
+            }
+        }
+
+        sim_path sp = calculatePath(steerPath);
+        cost = calculatePathCost(sp, kdtree);
+
+        weightTotal += 1.0 / cost;
+        nPathPoses++;
+
+        for(int j = 0; j < sp.angles.size(); j++) {
+            if(j < weightedSteeringRaw.size()) {
+                weightedSteeringRaw[j] += sp.angles[j] / cost;
+            } else {
+                weightedSteeringRaw.push_back(sp.angles[j] / cost);
+            }
+        }
+    }
+
+    vector<double> bestSteering;
+    for(int i = 0; i < weightedSteeringRaw.size(); i++) {
+        bestSteering.push_back(weightedSteeringRaw[i] / weightTotal);
+    }
+
+    sim_path bestPath = calculatePath(bestSteering);
+
+    desired_steer_angle = bestPath.angles[0];
+    desired_velocity = steeringToSpeed(desired_steer_angle);
```

Comment:
A small optimization, but isn't this speed already calculated at `bestPath.speeds[0]` ?